bug: Workflow controller crashes when missing ownerrefs

**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG
**What happened**:
Observed a panic due to missing ownerrefs for one Workflow.
```
{"level":"info","ts":1677874924.4694443,"logger":"addon-manager-wf-controller","msg":"reconciling","request":"addon-manager-system/prometheus-install-362c2de8-wf"," workflow ":"prometheus-install-362c2de8-wf"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1453607]

goroutine 436 [running]:
github.com/keikoproj/addon-manager/controllers.(*WorkflowReconciler).Reconcile(0xc000159780, {0x1977e38, 0xc000bb6450}, {{{0xc0012149c0, 0x16527e0}, {0xc001205fe0, 0xc00009e000}}})
	/workspace/controllers/workflow_controller.go:83 +0x4c7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00071ba40, {0x1977d90, 0xc000438b00}, {0x15e6a00, 0xc000fc6000})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:298 +0x303
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00071ba40, {0x1977d90, 0xc000438b00})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:214 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.2/pkg/internal/controller/controller.go:210 +0x354
```

```
kubectl get pods -n addon-manager-system
NAME                                                READY   STATUS             RESTARTS        AGE
addon-manager-controller-85c7489b46-k9cjh           1/2     CrashLoopBackOff   245 (49s ago)   14d
addon-manager-workflow-controller-66444744b-7ztvp   1/1     Running            0               14d
```

**What you expected to happen**:
* Panics should be caught and recovered.
* Figure out how Workflow can be missing ownerref?

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Addon Manager version
- Kubernetes version :
```
$ kubectl version -o yaml
```

**Other debugging information (if applicable)**:
- Addon status:
```
$ kubectl describe addon <addon-name>
```
- controller logs:
```
$ kubectl logs <addon-manager-pod>
```

--- a/controllers/workflow_controller.go
+++ b/controllers/workflow_controller.go
@@ -85,6 +85,11 @@ func (r *WorkflowReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	}
 
 	owner := metav1.GetControllerOf(wfobj)
+	if owner == nil {
+		err := fmt.Errorf("workflow %s/%s has no owner", wfobj.GetNamespace(), wfobj.GetName())
+		r.log.Error(err, wfobj.GetNamespace(), wfobj.GetName(), " owner is empty")
+		return ctrl.Result{}, err
+	}
 	if owner.Kind != "Addon" {
 		r.log.Info("workflow ", wfobj.GetNamespace(), wfobj.GetName(), " owner ", owner.Kind, " is not an addon")
 		return ctrl.Result{}, nil

--- a/controllers/workflow_controller.go
+++ b/controllers/workflow_controller.go
@@ -59,6 +59,11 @@ func NewWFController(mgr manager.Manager, dynClient dynamic.Interface, addonUpda
 // +kubebuilder:rbac:groups=argoproj.io,resources=workflows,namespace=system,verbs=get;list;watch;create;update;patch;delete
 
 func (r *WorkflowReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	defer func() {
+		if err := recover(); err != nil {
+			r.log.Info(fmt.Sprintf("Error: Panic occurred when reconciling %s due to %v", req.String(), err))
+		}
+	}()
 	wfobj := &wfv1.Workflow{}
 	err := r.client.Get(ctx, req.NamespacedName, wfobj)
 	if apierrors.IsNotFound(err) {

--- a/controllers/workflow_controller_test.go
+++ b/controllers/workflow_controller_test.go
@@ -6,6 +6,7 @@ import (
 
 	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 
 	logrtesting "github.com/go-logr/logr/testing"
@@ -34,20 +35,82 @@ func init() {
 func TestWorkflowReconciler_Reconcile(t *testing.T) {
 	g := gomega.NewGomegaWithT(t)
 
-	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithObjects().Build()
+	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).Build()
 	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
-	log := logrtesting.TestLogger{T: t}
-	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), log)
+	testLog := logrtesting.TestLogger{T: t}
+	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
 
 	r := &WorkflowReconciler{
 		client:       fakeCli,
 		dynClient:    dynFakeCli,
-		log:          log,
+		log:          testLog,
 		addonUpdater: addonUpdater,
 	}
 
 	res, err := r.Reconcile(ctx, controllerruntime.Request{
 		NamespacedName: types.NamespacedName{Namespace: "default", Name: "test"}})
 	g.Expect(err).To(gomega.BeNil())
+	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
+}
+
+func TestWorkflowReconciler_Reconcile_EmptyPhase(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	wf := &wfv1.Workflow{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "test",
+			Namespace:       "default",
+			OwnerReferences: nil,
+		},
+	}
+
+	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithRuntimeObjects(wf).Build()
+	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
+	testLog := logrtesting.TestLogger{T: t}
+	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
+
+	r := &WorkflowReconciler{
+		client:       fakeCli,
+		dynClient:    dynFakeCli,
+		log:          testLog,
+		addonUpdater: addonUpdater,
+	}
+
+	res, err := r.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{Namespace: "default", Name: "test"}})
+	g.Expect(err).To(gomega.BeNil())
+	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
+}
+
+func TestWorkflowReconciler_Reconcile_OwnerRefEmpty(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	wf := &wfv1.Workflow{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "test",
+			Namespace:       "default",
+			OwnerReferences: nil,
+		},
+		Status: wfv1.WorkflowStatus{
+			Phase: wfv1.WorkflowSucceeded,
+		},
+	}
+
+	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithRuntimeObjects(wf).Build()
+	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
+	testLog := logrtesting.TestLogger{T: t}
+	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
+
+	r := &WorkflowReconciler{
+		client:       fakeCli,
+		dynClient:    dynFakeCli,
+		log:          testLog,
+		addonUpdater: addonUpdater,
+	}
+
+	res, err := r.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{Namespace: "default", Name: "test"}})
+	g.Expect(err).To(gomega.HaveOccurred())
+	g.Expect(err).To(gomega.MatchError("workflow default/test has no owner"))
 	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
 }

--- a/controllers/workflow_controller_test.go
+++ b/controllers/workflow_controller_test.go
@@ -114,3 +114,37 @@ func TestWorkflowReconciler_Reconcile_OwnerRefEmpty(t *testing.T) {
 	g.Expect(err).To(gomega.MatchError("workflow default/test has no owner"))
 	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
 }
+
+func TestWorkflowReconciler_Reconcile_OwnerrefNotAddon(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	wf := &wfv1.Workflow{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: "default",
+			OwnerReferences: []metav1.OwnerReference{
+				{Kind: "Pod", Name: "test", UID: "test", APIVersion: "v1", Controller: &[]bool{true}[0]},
+			},
+		},
+		Status: wfv1.WorkflowStatus{
+			Phase: wfv1.WorkflowSucceeded,
+		},
+	}
+
+	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithRuntimeObjects(wf).Build()
+	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
+	testLog := logrtesting.TestLogger{T: t}
+	addonUpdater := pkgaddon.NewDefaultAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
+
+	r := &WorkflowReconciler{
+		client:       fakeCli,
+		dynClient:    dynFakeCli,
+		log:          testLog,
+		addonUpdater: addonUpdater,
+	}
+
+	res, err := r.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{Namespace: "default", Name: "test"}})
+	g.Expect(err).ToNot(gomega.HaveOccurred())
+	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
+}

--- a/controllers/workflow_controller_test.go
+++ b/controllers/workflow_controller_test.go
@@ -1,0 +1,53 @@
+package controllers
+
+import (
+	"testing"
+	"time"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	logrtesting "github.com/go-logr/logr/testing"
+	addonmgrv1alpha1 "github.com/keikoproj/addon-manager/api/addon/v1alpha1"
+	pkgaddon "github.com/keikoproj/addon-manager/pkg/addon"
+	"github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/runtime"
+	dynfake "k8s.io/client-go/dynamic/fake"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var (
+	testScheme = runtime.NewScheme()
+	rcdr       = record.NewBroadcasterForTests(1*time.Second).NewRecorder(testScheme, v1.EventSource{Component: "addons"})
+)
+
+func init() {
+	_ = addonmgrv1alpha1.AddToScheme(testScheme)
+	_ = wfv1.AddToScheme(testScheme)
+	_ = clientgoscheme.AddToScheme(testScheme)
+}
+
+func TestWorkflowReconciler_Reconcile(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithObjects().Build()
+	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
+	log := logrtesting.TestLogger{T: t}
+	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), log)
+
+	r := &WorkflowReconciler{
+		client:       fakeCli,
+		dynClient:    dynFakeCli,
+		log:          log,
+		addonUpdater: addonUpdater,
+	}
+
+	res, err := r.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{Namespace: "default", Name: "test"}})
+	g.Expect(err).To(gomega.BeNil())
+	g.Expect(res).To(gomega.Equal(controllerruntime.Result{}))
+}

--- a/controllers/workflow_controller_test.go
+++ b/controllers/workflow_controller_test.go
@@ -134,7 +134,7 @@ func TestWorkflowReconciler_Reconcile_OwnerrefNotAddon(t *testing.T) {
 	fakeCli := fake.NewClientBuilder().WithScheme(testScheme).WithRuntimeObjects(wf).Build()
 	dynFakeCli := dynfake.NewSimpleDynamicClient(testScheme)
 	testLog := logrtesting.TestLogger{T: t}
-	addonUpdater := pkgaddon.NewDefaultAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
+	addonUpdater := pkgaddon.NewAddonUpdater(rcdr, fakeCli, pkgaddon.NewAddonVersionCacheClient(), testLog)
 
 	r := &WorkflowReconciler{
 		client:       fakeCli,

